src/frontend/js/clusive-autosave.js

```diff
@@ -0,0 +1,69 @@
+/* Code for comprehension and affect assessments */
+/* global clusive, clusiveContext, PAGE_EVENT_ID, DJANGO_CSRF_TOKEN, fluid, D2Reader */
+/* exported clusiveAutosave */
+
+var clusiveAutosave = {
+    queue: clusive.djangoMessageQueue({
+            config: {                        
+                localStorageKey: "clusive.messageQueue.autosave",
+                lastQueueFlushInfoKey: "clusive.messageQueue.autosave.log.lastQueueFlushInfo"
+            }
+        }),
+    // Test if data is equivalent for autosave purposes        
+    isEquivalentData: function (oldData, newData) {
+        var isEquivalent = true;
+        Object.keys(newData).forEach(function (key) {
+            if(newData[key] !== oldData[key]) {
+                console.log("comparison of values", newData[key], oldData[key]);
+                console.log("found differing values for " + key, newData[key], oldData[key]);
+                isEquivalent = false;                
+            }
+        });
+        console.log("isEquivalent", isEquivalent);
+        return isEquivalent;                
+    },
+    save: function(url, data) {        
+        var lastData = clusiveAutosave.lastDataCache[url];
+        // TODO: Only check that same keys with same values exist
+        var isNewData = !clusiveAutosave.isEquivalentData(lastData, data);                
+        if(isNewData) {
+            console.log("adding changed data to autosave queue");
+            clusiveAutosave.queue.add({"type": "AS", "url": url, "data": JSON.stringify(data)});
+            clusiveAutosave.lastDataCache[url] = data; 
+        } else {
+            console.log("data has not changed, not adding to autosave queue");
+        }
+    },
+    retrieve: 
+        function(url, callback) {
+            var hasLocal = false;                
```

Comment:
The `hasLocal` variable should actually be removed - checking for local data is already functionally accomplished by the `autoSaveMessages` test.

Since user interaction in another tab / window could have updated the results, every autosave based retrieval needs to go through the message queue to make sure it's got the latest data. `lastDataCache` is just to make the autosave doesn't push messages into the queue unnecessarily.